ENH: added NMIT method

nbokulich opened the pull request:
fixes #53 

This method generates a distance matrix based on temporal similarity between individuals sampled longitudinally. Pass to `beta-group-significance` or `emperor` to compare/view sample similarities.

Very many thanks to @elong0527 who wrote most of the core nmit utility functions.

The `ecam_map_small.txt` and `ecam-table-small.qza` files are just new tiny test files (NMIT can take a long time to run) and don't really need review.

nbokulich committed:
nmit method

nbokulich committed:
nmit utilities tests

nbokulich committed:
nmit tests

nbokulich committed:
tutorial sample md

elong0527 committed:
add id_set option in _temporal_corr

The id_set is required to make sure the order of the correlation keep the same while calculate the distance matrix.

nbokulich committed:
Merge pull request #1 from elong0527/patch-1

add id_set option in _temporal_corr

nbokulich committed:
cleaning up tests

nbokulich commented on q2_longitudinal/_utilities.py:
```diff
@@ -577,3 +578,74 @@ def _stats_and_visuals(output_dir, pairs, metric, group_column,
+    id_n = len(id_set)
+
+    dist = np.zeros((id_n, id_n))
+    for i, id_i in enumerate(id_set):
```
@elong0527 it looks like there may be two unnecessary steps in this loop:
1. this will calculate the distance of each `(id_i, id_j)` pair twice. we should instead just duplicate the result if this creates a symmetric matrix
2. this calculates the identical-sample distance along the diagonal of the matrix. Should this just be skipped so that all self-comparisons are 0?

elong0527 replied:
Yes, you are right. I will update the code to improve performance.

thermokarst replied:
How about something like this?

```python
dist = np.zeros((id_n, id_n))
for i, id_i in enumerate(id_set):
    for j, id_j in enumerate(id_set[:i]):
        dist[i, j] = dist[j, i] = linalg.norm(corr.loc[id_i] - corr.loc[id_j], ord=dist_method)
```

elong0527 committed:
Update nmit corrlation matrix

Increase effciency by using pandas rountine function.

elong0527 committed:
test push function

No change of the code.

I hope to understand how to push multiple commits into one pull request.

elong0527 committed:
Revert "test push function"

This reverts commit b12e6bdb17ced43ee0d0437561d27aa6a7988e46.

elong0527 committed:
Update formating

Update formating by using flake8.

elong0527 committed:
Add _temporal_screening

Add taxa screening  function for nmit

nbokulich committed:
fixed unit tests

nbokulich committed:
fixed unit tests (#4)

* Update nmit corrlation matrix

Increase effciency by using pandas rountine function.

* test push function

No change of the code.

I hope to understand how to push multiple commits into one pull request.

* Revert "test push function"

This reverts commit b12e6bdb17ced43ee0d0437561d27aa6a7988e46.

* Update formating

Update formating by using flake8.

* Add _temporal_screening

Add taxa screening  function for nmit

* fixed unit tests

nbokulich committed:
smaller taxa table for testing in tutorial data

thermokarst commented on q2_longitudinal/tests/data/ecam_map_small.txt:
```diff
@@ -0,0 +1,19 @@
+#SampleID	antiexposedall	day_of_life	delivery	diet	diet_3	mom_child	month	month_of_life	sample_summary	sex	studyid
```
I don't see this file being utilized anywhere in the tests - is it necessary? Or maybe a test wasn't updated that should be using this?

nbokulich replied:
~~See line [197](https://github.com/nbokulich/q2-intervention/blob/issue-53/q2_longitudinal/tests/test_longitudinal.py#L197) and [282](https://github.com/nbokulich/q2-intervention/blob/issue-53/q2_longitudinal/tests/test_longitudinal.py#L282)~~

thermokarst replied:
Hmm, those lines aren't referencing this file, sorry if it is right in front of me...

nbokulich replied:
you are right, I have removed this file. thanks!

thermokarst commented on tutorial_data/ecam_map_small.txt:
```diff
@@ -0,0 +1,19 @@
+#SampleID	antiexposedall	day_of_life	delivery	diet	diet_3	mom_child	month	month_of_life	sample_summary	sex	studyid
```
Sorry I didn't notice this before, but do we need the `tutorial_data/` directory? Seems like now that these files are up on data.qiime2.org we should remove this from the repo. Thoughts?

nbokulich replied:
no — I suppose it is safe to remove this now. Will remove in this round of changes.

thermokarst commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -172,3 +172,24 @@ def volatility(output_dir: str, metadata: qiime2.Metadata, group_column: str,
+         individual_id_column: str, corr_method="kendall", dist_method="fro"
+         ) -> DistanceMatrix:
+
+    # load and merge feature table and metadata to ensure IDs match
```
This sounds like we want to ensure that the biom table ids are a subset of the metadata ids, then, only process the intersection of the two (as in, the biom ids). We have an example of this kind of validation [here](https://github.com/qiime2/q2-feature-table/blob/master/q2_feature_table/_heatmap/_visualizer.py#L78-L82), if you are interested.

nbokulich replied:
~~we want to make sure that the biom and metadata have identical index (sample IDs) — so filtering (of both dataframes), not validation.~~

nbokulich replied:
👍

thermokarst commented on q2_longitudinal/_longitudinal.py:
```diff
@@ -172,3 +172,24 @@ def volatility(output_dir: str, metadata: qiime2.Metadata, group_column: str,
+    # load and merge feature table and metadata to ensure IDs match
+    metadata = _load_metadata(metadata)
+    metadata = metadata[[individual_id_column]]
+    metadata = pd.concat([metadata, table], axis=1, join='inner')
```
Personal preference here, I prefer to use [`pd.join`](https://github.com/qiime2/q2-feature-table/blob/master/q2_feature_table/_heatmap/_visualizer.py#L94). I like to use this because it borrows from database terminology, which is arguably a little more explicit than `concat`ing and various axes (a join is always a row-wise join in db-land). No need to change, just wanted to put it out there.

nbokulich replied:
I like that too — it's shorter and more elegant than what I have, but it's also causing something to break downstream. So if this is just a matter of style, I'll leave what i have for now.

thermokarst commented on q2_longitudinal/_utilities.py:
```diff
@@ -577,3 +578,78 @@ def _stats_and_visuals(output_dir, pairs, metric, group_column,
+def _temporal_corr(taxa, individual_id, corr_method="kendall"):
+    '''Create Temporal correlation from a feature table
+    containing repeated measures samples.
+    taxa: pd.DataFrame
```
It is a little strange to refer to this table as `taxa` here, maybe we could rename to `table`?

Does this method make any special assumptions about feature identifiers in the table? Is it assuming they will be taxonomic identifiers? If so, we should probably chat about this a bit.

nbokulich replied:
👍

elong0527 replied:
I made the change, it is `table` originally, I am fine to use taxa/table to call it. Just to remind the user this is a OTU table :)

thermokarst commented on q2_longitudinal/_utilities.py:
```diff
@@ -577,3 +578,78 @@ def _stats_and_visuals(output_dir, pairs, metric, group_column,
+
+def _temporal_distance(corr, id_set, dist_method="fro"):
+    '''Calculate Distance Matrix from temporal correlation data.
+    corr_dict: pd.DataFrame
```
This docstring doesn't match the signature.

nbokulich replied:
thanks for catching! have checked all docstrings.

thermokarst commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -28,10 +28,14 @@
 
 
-base_parameters = {
+basest_parameters = {
```
Yowzah! How about `shared_parameters`?

nbokulich replied:
😞 I guess I was just channeling a little shakespeare or something... "shared" it is!

thermokarst commented on q2_longitudinal/plugin_setup.py:
```diff
@@ -47,26 +51,32 @@
+            'either state_1 or state_2, that subject will be dropped and '
+            'reported in standard output by default. Set '
+            'replicate_handling="random" to instead randomly select one '
+            'member, and use --verbose to list conflicts.'),
```
`--verbose` is a cli-specific flag, but this method can be run in any of the QIIME 2 interfaces. Perhaps just 

> and run in verbose-mode to list conflicts.

Or something to that end.

nbokulich replied:
Thanks for catching. That's actually an old description I should have removed earlier. In actions where replicate handling occurs, a warning now appears in the viz to list samples that were dropped.

thermokarst commented on q2_longitudinal/tests/test_longitudinal.py:
```diff
@@ -171,6 +194,7 @@ def _load_dm(dm_fp):
             return dm
 
         self.table_ecam_fp = _load_features('ecam-table-maturity.qza')
+        self.table_taxa_fp = _load_features('ecam-table-small.qza')
```
~see above re: `taxa` vs `table`~

thermokarst reviewed (changes requested):
Thanks @nbokulich! I have a handful of questions and comments - please let me know if you need any clarification or help! Thanks!

nbokulich committed:
Merge branch 'issue-53' of https://github.com/nbokulich/q2-intervention into issue-53-test

merge upstream changes

nbokulich committed:
review corrections

nbokulich committed:
selecting by index is tidier than concat/drop

thermokarst committed:
SQUASH:

- Replace OTU table with Feature Table
- Remove and update comments
- Dropped unnecessary metadata munging
- Added missing test for ensuring table ids are covered by MD

thermokarst committed:
SQUASH

- Remove `species` from correlation comment
